src/fe/fe.C

```diff
@@ -58,10 +58,10 @@ FE<Dim,T>::FE (const FEType & fet) :
 
 
 template <unsigned int Dim, FEFamily T>
-unsigned int FE<Dim,T>::n_shape_functions () const
+unsigned int FE<Dim,T>::n_shape_functions (const bool add_p_level) const
 {
   return FE<Dim,T>::n_dofs (this->elem_type,
-                            static_cast<Order>(this->fe_type.order + this->_p_level));
+                            static_cast<Order>(this->fe_type.order + add_p_level * this->_p_level));
```

Comment:
This seems weird to me.  The premise of caching `elem_type` and `_p_level` is that they should represent whatever we did our last `reinit()` with, isn't it?  So we should never have to override the `n_shape_functions()` behavior, because "return the number of shape functions we would have had if we had done a thing we didn't do" isn't a thing anyone needs to know, is it?

It looks to me like we're setting `this->_p_level = elem->p_level()` even when `this->_add_p_level_in_reinit` is false.  Isn't that a bug, and if we fixed that would it obviate the need for this PR?

Reply:
Do you regard `get_order` as cached data as well? If so, I will fix that API as well

Reply:
I would suppose yes since it's using `this->_p_level`